current 24.0rc testing guide (#1)
kouloumos/rc24-guide-feedback-repo: testing_guide.md

# Testing Guide: Bitcoin Core 24.0 Release Candidate

_For feedback on this guide, please visit [#26092](https://github.com/bitcoin/bitcoin/issues/26092)_

This document outlines some of the upcoming Bitcoin Core 24.0 release changes and provides steps to help test them. This guide is meant to be the starting point for experimentation and further testing, but is in no way comprehensive! After running through the steps in this guide, you are encouraged to do your own testing.

> This can be as simple as testing the same features in this guide but trying it a different way. Even better, think of features you use regularly and test that they still work as expected in the release candidate. You can also read the [release notes][release notes] to find something not covered in this guide. **This is a great way to be involved with Bitcoin's development and helps keep Bitcoin running smoothly and bug-free! Your help in this endeavor is greatly appreciated.**

Reviewer: testing out the same features in a different way seems confusing, how am i going to test out the gui restore from wallet in a different way?

maybe something like, "This can be as simple as testing that the new release finishes syncing on your machine"?

Author: Later in the guide I've tried to add some examples of "testing it in a different way" to help people think ideas.

For example, for the GUI restore option: 
> A slightly different testing scenario is to [stop your node and clean the datadir](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/24.0-Release-Candidate-Testing-Guide#4-reset-testing-environment) between backup and restore. Or maybe move the backup file and restore to a different machine.

Or for the unification of settings - talking about pruning settings:
> A different test is to enforce pruning by adding prune=x in bitcoin.conf, then change it in the GUI and verify that the GUI option takes precedence in bitcoind over what's set in bitcoin.conf.

## Overview

Changes covered in this testing guide include:

> - Observing the new headers pre-synchronization phase during IBD ([#25717](https://github.com/bitcoin/bitcoin/pull/25717))

Reviewer: maybe add anti-dos here and other places where this is referenced?

## Preparation

### 1. Grab Latest Release Candidate

**Current Release Candidate:** [Bitcoin Core 23.0rc1][source-code] ([release-notes][release notes])

There are two ways to grab the latest release candidate: pre-compiled binary or source code. The source code for the latest release can be grabbed from here: [latest release source code][source-code]

If you want to use a binary, make sure to grab the correct one for your system. [NA]

[source-code]: https://github.com/bitcoin/bitcoin/releases/tag/v24.0rc1

### 2. Compile Release Candidate

**If you grabbed a binary, skip this step**

Before compiling, make sure that your system has all the right dependencies installed. As this guide utilizes the Bitcoin Core GUI, you must compile support for the GUI and have the `qt5` dependency already installed. Since descriptor wallets are now the default, it is recommended that you compile with sqlite, so make sure you have installed the `sqlite3` dependency.

To ensure we build in sqlite and gui support, pass the following options when configuring:

```shell
./autogen.sh
./configure --with-gui=yes --with-sqlite=yes --enable-suppress-external-warnings

> # nproc does not exist on macOS so we create an alias for it if necessary

Reviewer: Remove the alias and note on actual building from the section on compiling the release candidate.

The note added here is really a hint towards the configuration we want to do, but it is not complete. These steps as outlined won't work for all systems, there's extra configuration needed say for example on the bsd's. Some user's won't actually be able to build with these steps.

So, let the user read the build docs if they're going to go on the adventure of compiling core, let them bathe in the build system and be reborn

Author: good point.

For more information on compiling from source, here are some guides to compile Bitcoin Core for [UNIX/Linux](https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md), [macOS](https://github.com/bitcoin/bitcoin/blob/master/doc/build-osx.md), [Windows](https://github.com/bitcoin/bitcoin/blob/master/doc/build-windows.md), [FreeBSD](https://github.com/bitcoin/bitcoin/blob/master/doc/build-freebsd.md), [NetBSD](https://github.com/bitcoin/bitcoin/blob/master/doc/build-netbsd.md), and [OpenBSD](https://github.com/bitcoin/bitcoin/blob/master/doc/build-openbsd.md).

### 3. Setting up command line environment

If you plan to use the command line, below are a few environment variable to set.

First, create a temporary data directory

```shell
export DATA_DIR=/tmp/24-rc-test
mkdir $DATA_DIR
```

Next, specify the following paths. For source compiled, start from the root of your release candidate directory and run:

```shell
export BINARY_PATH=$(pwd)/src
export QT_PATH=$(pwd)/src/qt
```

For the downloaded binary, start from the root of the downloaded release candidate (cd ~/bitcoin-24.0rc1, for example) and run:

```shell
export BINARY_PATH=$(pwd)/bin
export QT_PATH=$BINARY_PATH
```

> To avoid specifying the data directory (`-datadir=$DATA_DIR`) on each command, bellow are a few extra variables to set.

Reviewer: ```suggestion
To avoid specifying the data directory (`-datadir=$DATA_DIR`) on each command, bellow are a few extra variables to set.
```
```suggestion
To avoid specifying the data directory (`-datadir=$DATA_DIR`) on each command, below are a few extra variables to set.
```

> :information_source: **Note:** If you are testing on signet, you can also include the `-signet` flag

```shell
alias bitcoind="$(echo $BINARY_PATH)/bitcoind -datadir=$DATA_DIR"
alias cli="$(echo $BINARY_PATH)/bitcoin-cli -datadir=$DATA_DIR"
alias qt="$(echo $QT_PATH)/bitcoin-qt -datadir=$DATA_DIR"
```

The commands throughout the rest of the guide will look like:

```shell
cli [cli args]

# for starting bitcoin-qt
qt [cli args]
```

### 4. Reset testing environment

> Between sections in this guide, it's recommended to stop your node and wipe the data directory. You can use the commands provided below.

Reviewer: could this be aliased?

Author: good point!

Author: I was expecting that commands could be chained but in practise this creates some concurrency issues because $DATA_DIR is deleted before shutdown is completed.

**Stop node**

```shell
cli stop
```

**Wipe and recreate the directory**

```shell
rm -r $DATA_DIR
mkdir $DATA_DIR
```

**Start node**

```shell
bitcoind -daemon
```

> ### 5. Testing with a signet faucet

Reviewer: why not point to the [signet/getcoins.py](https://github.com/bitcoin/bitcoin/blob/master/contrib/signet/getcoins.py) script?

Author: good point, I'll do. Although pointing to the script was part of the 22.0 testing-guide, it wasn't on the next one and I dismissed it.

If you are testing on signet, you can use test coins for the tests in this guide. After IBD has completed (this process is relatively quick on signet), you can get some free corn from one of the signet faucets available, e.g. https://signetfaucet.com/. Generate a fresh address when needed, pop it into the faucet, and [watch the signet coins flowing in](https://mempool.space/signet).

When you have finished testing, please [return your coins to a faucet in need](https://signet.bublina.eu.org/about.html) before stopping your node and cleaning up your datadir.

> :bulb: An efficent way to sweep your testing wallet is the new `sendall` RPC ([#24118](https://github.com/bitcoin/bitcoin/pull/24118)).

### 6. Testing using QT

> If you're not comfortable with the command line, you can still test all of these changes in the GUI. Althought for some steps you might need to use the integrated RPC console (Window->Console).

Reviewer: ```suggestion
If you're not comfortable with the command line, you can still test all of these changes in the GUI. Although, for some steps, you will need to use the integrated RPC console (Window->Console).
```

**The logic for downloading headers from peers has been reworked to address this potential denial-of-service**. This is particularly relevant for nodes starting up for the first time (or for nodes which are starting up after being offline for a long time).

With this new logic, the node downloads headers from a given peer twice:

1. Download (Pre-synchronizing) blockheaders: to verify (using `nMinimumChainWork`) that the header is part of a chain that has sufficiently high work, prior to storing those headers permanently.
2. Redownload (Synchronizing) blockheaders: to fully validate and permanently store the headers.

> :book: **_What is `nMinimumChainWork`?_** A number designed to protect new clients from accepting fake blockchains during initial sync. It's [updated on every release](https://github.com/bitcoin/bitcoin/pull/25946) and defines the minimum amount of total work a chain must have before the client considers it valid.

For this test, we will start by observing the pre-syncing phase during IBD and then verify that the new node is actually able to sync. This can be tested with or without GUI.

### Starting IBD

> You can observe the pre-synchronization phase in detail by using the log file. With or without GUI, the logs are located at `$DATA_DIR/debug.log`. For the relevant logs to appear in your `debug.log` file, you'll need to use the `-debug` flag with the `net` category. If you are using bitcoind you'll also be able to observe real-time logging from the console.

Reviewer: ```suggestion
You can observe the pre-synchronization phase in detail by using the log file. With or without the GUI, the logs are located at `$DATA_DIR/debug.log`. For the relevant logs to appear in your `debug.log` file, you'll need to use the `-debug` flag with the `net` category. If you are using bitcoind you'll also be able to observe real-time logging from the console.
```

Reviewer: You could also see real-time logging from the terminal if someone is running the gui and passes -printtoconsole

Author: Nice, I'll add it as a tip.

> For the first part of this test, there is no need to sync the full blockchain. Therefore, you can use the `-stopatheight` flag, for the process to terminate almost as soon as the headers synchronization phase (download&redownload) is over.

Reviewer: ```suggestion
For the first part of this test, there is no need to sync the full blockchain. Therefore, you can use the `-stopatheight` flag, for the process to terminate almost as soon as the headers synchronization phase (download & re-download) is over.
```

2022-09-18T10:08:38Z [net] disconnecting peer=0
[...]
2022-09-18T10:08:45Z [net] Initial headers sync started with peer=2: height=0, max_commitments=4443456, min_work=00000000000000000000000000000000000000003404ba0801921119f903495e
2022-09-18T10:08:45Z [net] sending getheaders (101 bytes) peer=2
2022-09-18T10:08:45Z [net] more getheaders (from 00000000dfd5d65c9d8561b4b8f60a63018fe3933ecb131fb37f905f87da951a) to peer=2
2022-09-18T10:08:45Z Pre-synchronizing blockheaders, height: 2000 (~0.28%)
```

Being in IBD doesn't stop other peers from announcing new blocks (every ~10 minutes) to us. When a node receives a block announcement, **it used to add all announcing peers as additional headers sync peers** and initiate headers sync. The addition of the pre-sync phase with [PR#25717](https://github.com/bitcoin/bitcoin/pull/25717) made this behavior more bandwidth-wasteful.

In an effort to strike a better balance between robustness in the face of stalling peers and bandwidth usage, [PR#25720](https://github.com/bitcoin/bitcoin/pull/25720) changed this logic such that **now only one of the announcing peers is added for headers sync** (one sync for each new block announcement).

> **You can observe this change in behavior if a new block is found during the blockheaders synchronization**. Bellow, during header sync with peer 2, we receive a block announcement (via an `inv` message) from peer 3. Observe that from that point onwards we request the same headers from both peers.

Reviewer: https://www.merriam-webster.com/dictionary/bellow
```suggestion
**You can observe this change in behavior if a new block is found during the blockheaders synchronization**. Below, during header sync with peer 2, we receive a block announcement (via an `inv` message) from peer 3. Observe that from that point onwards we request the same headers from both peers.
```

Author: TIL.

Below is an example showing a conflict for the `Alt+R` shortcut in the Spanish GUI. Observe that by pressing `Alt`, shortcut hints are displayed under the `<Key>` of each option. **The key `R` is part of three different options, resulting in ambiguity**.

<img width="450" alt="gui shortcut ambiguity" src="https://user-images.githubusercontent.com/18506343/191549555-afcfd791-e21c-47ea-b63e-38b34cee0cdb.gif">

This is a pretty simple test - **navigate in the GUI, use the shortcuts and see if you can find any more ambiguities**.

This kind of issues are responsibility of the translators. If you want to directly help with a fix, see [doc/translation_process.md](https://github.com/bitcoin-core/gui/blob/master/doc/translation_process.md). The best way to report this is to open an issue on [Bitcoin Core's Transifex page](https://www.transifex.com/bitcoin/bitcoin/), see how on [Comments and Issues section on Transifex docs](https://docs.transifex.com/translation/tools-in-the-editor#comments-and-issues). Alternatively, comment [here](https://github.com/bitcoin-core/gui/issues/126) with details.

### Testing the Restore Wallet option

> Restoring a wallet from a backup file in the GUI is now possible! Up until now, you could create a backup but the only option was to restore using the RPC interface. This change makes it easier for non-technical users to restore backups.

Reviewer: if it's someone first foray into using the gui, it might not be apparent how you would go about creating a backup, You should atleast be explicit about the menu options that need to be clicked in order to accomplish this test

Author: Good point, I'll add the menu options sequence.

- Testing transient addresses for I2P outbound connections ([#25355](https://github.com/bitcoin/bitcoin/pull/25355))
- Using the new `migratewallet` RPC to migrate legacy wallets to descriptor wallets. ([#19602](https://github.com/bitcoin/bitcoin/pull/19602))
- Testing watch-only support for Miniscript descriptors. ([#24148](https://github.com/bitcoin/bitcoin/pull/24148))

For a comprehensive list of changes in Bitcoin Core 24.0, check out the [release notes][release notes].

## Preparation

### 1. Grab Latest Release Candidate

**Current Release Candidate:** [Bitcoin Core 24.0rc1][source-code] ([release-notes][release notes])

There are two ways to grab the latest release candidate: pre-compiled binary or source code. The source code for the latest release can be grabbed from here: [latest release source code][source-code]

If you want to use a binary, make sure to [grab the correct one for your system][binaries].

[source-code]: https://github.com/bitcoin/bitcoin/releases/tag/v24.0rc1
[binaries]: https://bitcoincore.org/bin/bitcoin-core-24.0/test.rc1/

### 2. Compile Release Candidate

**If you grabbed a binary, skip this step**

> Before compiling, make sure that your system has all the right dependencies installed. As this guide utilizes the Bitcoin Core GUI, you must **compile with support for the GUI**(`--with-gui=yes`) and have the `qt5` dependency already installed. Since descriptor wallets are now the default, it is recommended that you **compile with sqlite**(`--with-sqlite=yes`), so make sure you have installed the `sqlite3` dependency.

Reviewer 2: On the topic of having the sqlite and qt dependencies already installed, it may be worthwhile to leave a note that you will need to compile with berkeleydb if you plan on testing any legacy wallet migration stuff from scratch. 

~~I found that the [openbsd installation docs](https://github.com/bitcoin/bitcoin/blob/master/doc/build-openbsd.md#3-install-optional-dependencies) were the most useful for installing it, as opposed to the link in the [dependencies](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md) doc that goes straight to Oracle (and requires registration with Oracle)~~

EDIT: crossed out that second line, which is resolved in the comments below

Reviewer: The issue would be on many linux distros where the bdb version we want isn’t available. You could install a new version and pass ‘with-incompatible-bdb’ to configure. Alternatively, there’s a helper script to grab and compile the bdb we want.

but there’s an easy solution to any user configuration issues, point them to our depends build system :D

Reviewer 2: omg, I was clicking the link in the "Releases" column (which takes me to the Oracle site) instead of the link in the  "Dependency" column (which takes me to the helper script) in the [dependencies](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md) doc, and I somehow missed it on the installation doc for my distro. Totally agree on just pointing to the depends build system 👍

Author: Mentioning BDB for legacy wallets is indeed useful, I've just added that.

### What to expect after the migration

A new descriptor wallet creates 8 descriptors by default, 2 (external, internal) for each supported script type:

- P2PKH: `pkh([<fingerprint>/44'/1'/0']<xpub>/{0,1}/*)#<checksum>`
- P2SH(P2WPKH): `sh(wpkh([<fingerprint>/49'/1'/0']<xpub>/{0,1}/*))#<checksum>`
- P2WPKH: `wpkh([<fingerprint>/84'/1'/0']<xpub>/{0,1}/*)#<checksum>`
- P2TR: `tr([<fingerprint>/86'/1'/0']<xpub>/{0,1}/*)#<checksum>`

> :information_source: You can verify that with the `listdescriptors` RPC on a new descriptor wallet. For more information on output descriptors, see [doc/descriptors.md][descriptors.md].

A legacy wallet with an HD chain has an HD seed. This seed is hashed according to the BIP32 specification to become the BIP32 master key which everything else is then derived from. We need to associate those addresses with a descriptor. Because of that, in addition to the above, the new migrant wallet is expected to have:

- A range combo descriptor for the external addresses: `combo(<xpub>/0'/0'/*')#<checksum>`
- A range combo descriptor for the internal addresses: `combo(<xpub>/0'/1'/*')#<checksum>`

> - A combo descriptor for the hd seed: `combo(<hdseed>)#<checksum>`

Reviewer 2: I believe this should be two combo descriptors for the hd seed, right? One for internal (change) and one for external (receive)? I ran through this test scenario [twice](https://github.com/satsie/bitcoin-notez/blob/master/v24.0rc1-testing.md#migrating-legacy-wallets-to-descriptor-wallets) and both times I got two combo descriptors for the hd seed.

Author: I've read Will's answer on IRC and I thought that you cleared this out, just repeating it here to make sure. 

The testing guide says:
> For wallets containing non-HD keys, each key will have its own combo descriptor.

That's the [non-hd keys section](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/24.0-Release-Candidate-Testing-Guide/#non-hd-keys). There, we are using the `importprivkey` RPC. That's where that second combo descriptor comes from, it's not internal/external. One is from the hd seed and one is for the non-hd key. 

I've also have a note at the what to expect section that says
> ℹ️ The HD seed is a valid key that could receive Bitcoin to it even though its corresponding addresses would never be given out.

Essentially that means that the HD seed is another non-hd key.

Please let me know if how it is stated in the testing guide creates confusion.

Author: I've made a small modification at that part to try and make it a bit more clear.

Reviewer 2: ah!! I understand now :) Thanks for clarifying. I totally forgot the test has you import a non-HD key. It was a lot of steps for someone like me that has zero idea what they are doing haha :sweat_smile:. It makes sense that this would be the reason for the second combo descriptor. In general I'm not used to working with wallets that have non-HD keys imported in, so it was a new paradigm for me to wrap my head around.

I think at the end of the migration, when I scrolled back up to the "What to expect after the migration" section I assumed this part was going to be a checklist:

- A range combo descriptor for the external addresses: `combo(<xpub>/0'/0'/*')#<checksum>`
-  A range combo descriptor for the internal addresses: `combo(<xpub>/0'/1'/*')#<checksum>`
-  A combo descriptor for the hd seed: `combo(<hdseed>)#<checksum>`

And for the most part, it is. I just got so hung up on the last bit where I had two combo descriptors, I didn't try to read more into the stuff immediately below about the non-HD keys. 

Reflecting on it now, the wording of:

> For wallets containing non-HD keys, the new migrant wallet is expected... 

probably also threw me off. I followed all the steps in the test guide, so the part about importing the non-HD key didn't seem optional to me, yet this statement felt like I only needed to be concerned if I did something advanced or special.

Anyways, all of this is to say I think only extremely new readers will find themselves confused the way I did, and in general I'd say it's a non-issue. Certainly not worth splitting hairs over!

This concludes my feedback on the guide. Thank you for all the work you put into developing it, hosting PR review club, and sticking around to answer questions. As a beginner it was so wonderful to have a guide that walked me through all the steps, and taught me about different facets of the system I knew nothing about. Glad I was able to do some testing on the upcoming release, and that wouldn't have been possible for me without this guide! :star2:

> :information_source: The HD seed is a valid key that could receive Bitcoin to it even though it's corresponding addresses would never be given out.

For wallets containing non-HD keys, each key will have its own combo descriptor.

After a successful migration, all the above descriptors will be part of a newly created Descriptor wallet named as the original wallet.

### Preparing a Legacy wallet

Create a legacy test-wallet. This is the wallet we will migrate [later](#migrating-a-legacy-wallet) in the guide.

```shell

> cli -named createwallet wallet_name=legacy descriptors=false

Reviewer 2: When I try to run this I get the following error:

```
error code: -4
error message:
Compiled without bdb support (required for legacy wallets)
```

During initial setup at the beginning of the doc, I used this command to configure: `./configure --with-gui=yes --with-sqlite=yes --enable-suppress-external-warnings`

When I run that, I see that bdb is actually disabled:
```
Options used to compile and link:
  external signer = yes
  [...]
  with wallet     = yes
    with sqlite   = yes
    with bdb      = no
```

I see there is a `--without-bdb` configure flag, but I can't find any kind of "with-bdb" flag. `configure.ac` has me to believe that bdb should be enabled by default. 

Is there a way to override this and get bdb enabled?

Reviewer: You don’t have the bdb dependency installed 🚔🚨🚔

Reviewer 2: doh! :woman_facepalming:  thank you! I'll test this later tonight. It will probably lead to a suggestion to add this to the Preparation section, but I will wait till I can confirm my findings. :+1:

Reviewer 2: It works! I just left a [comment](https://github.com/kouloumos/rc24-guide-feedback-repo/pull/1/files#r982986596) in an earlier section (same area that sqlite and qt5 are mentioned) about giving the user a heads up that they will need berkeleydb installed. Feel free to resolve this one!

Once the test is successful, don't forget to [stop your node and clean the datadir](#4-reset-testing-environment).

## The Most Important Step

Thank you for your help in making Bitcoin as robust as it can be. Please remember to add a comment on [v24.0 testing issue detailing](https://github.com/bitcoin/bitcoin/issues/26175):

1. Your hardware and operating system
2. Which release candidate you tested and whether you compiled from source or used a binary (e.g. 24.0rc1 binary or 24.0rc1 compiled from source)
3. What you tested
4. Any other relevant findings

Don't be shy about leaving a comment even if everything worked as expected! We want to hear from you and so **it doesn't count unless you leave some feedback**.

Thanks for your contribution and for taking the time to make Bitcoin awesome. For feedback on this guide, please visit [#26092](https://github.com/bitcoin/bitcoin/issues/26092)

[release notes]: https://github.com/bitcoin-core/bitcoin-devwiki/wiki/24.0-Release-Notes-draft
[descriptors.md]: https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md
[miniscript homepage]: https://bitcoin.sipa.be/miniscript/
[miniscript.fun]: https://miniscript.fun/

> [miniscript-example]: https://miniscript.fun/#/full/eyJpZCI6ImRlbW9AMC4xLjAiLCJub2RlcyI6eyIyIjp7ImlkIjoyLCJkYXRhIjp7Im1uZW1vbmljIjoiY3Jhc2ggZmF0YWwgaG9sbG93IHRoYW5rIHN3YWxsb3cgc3VibWl0IHRhdHRvbyBwb3J0aW9uIGNvZGUgZm9hbSBtYXRoIGZvcmNlIiwicGFzc3dvcmQiOiIiLCJkZXJpdmF0aW9uIjoiIn0sImlucHV0cyI6e30sIm91dHB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjozLCJpbnB1dCI6ImtleSIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlstNjgyLjk3ODA3NzAxNzExNTUsLTk0LjM1NDg3NDg2ODg2NzM4XSwibmFtZSI6IkJJUDM5In0sIjMiOnsiaWQiOjMsImRhdGEiOnt9LCJpbnB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoyLCJvdXRwdXQiOiJrZXkiLCJkYXRhIjp7fX1dfX0sIm91dHB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbXX19LCJwb3NpdGlvbiI6Wy0zNTQuMzg1MjYxNjAxMzIyNiwtMjYuMTcwMDU4OTc5NTg0MzIyXSwibmFtZSI6IktleSJ9LCI2Ijp7ImlkIjo2LCJkYXRhIjp7fSwiaW5wdXRzIjp7InBvbCI6eyJjb25uZWN0aW9ucyI6W3sibm9kZSI6MTcsIm91dHB1dCI6InBvbCIsImRhdGEiOnt9fV19fSwib3V0cHV0cyI6eyJkZXNjIjp7ImNvbm5lY3Rpb25zIjpbXX19LCJwb3NpdGlvbiI6WzU1Ni45NTE5NTcxNzE2NDE0LDEyNC42MTE0ODQ0NTUzODg2Ml0sIm5hbWUiOiJEZXNjcmlwdG9yIn0sIjgiOnsiaWQiOjgsImRhdGEiOnsibnVtIjoyMX0sImlucHV0cyI6e30sIm91dHB1dHMiOnsicG9sIjp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoxNywiaW5wdXQiOiJwb2wxIiwiZGF0YSI6e319XX19LCJwb3NpdGlvbiI6Wy03OC41NTk4NTE4NjQzMzg0OCwtNjUuNTQ3MDM1NjYxMDU4NDZdLCJuYW1lIjoiQWZ0ZXIifSwiMTAiOnsiaWQiOjEwLCJkYXRhIjp7fSwiaW5wdXRzIjp7ImtleSI6eyJjb25uZWN0aW9ucyI6W3sibm9kZSI6MTIsIm91dHB1dCI6ImtleSIsImRhdGEiOnt9fV19fSwib3V0cHV0cyI6eyJrZXkiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjE2LCJpbnB1dCI6InBvbGljaWVzIiwiZGF0YSI6e319XX19LCJwb3NpdGlvbiI6Wy0zNDEuMjAzMDYwNTIzMDI2Myw0OTYuNTM3NjQ5NTU1NjA3NF0sIm5hbWUiOiJLZXkifSwiMTIiOnsiaWQiOjEyLCJkYXRhIjp7Im1uZW1vbmljIjoic3BlbGwgYm9yZGVyIGV4aGF1c3QgeWFyZCBlY29ub215IGZvc3NpbCBhcm15IGd1ZXNzIHR5cGljYWwgYm9tYiBhY2hpZXZlIGFtb3VudCJ9LCJpbnB1dHMiOnt9LCJvdXRwdXRzIjp7ImtleSI6eyJjb25uZWN0aW9ucyI6W3sibm9kZSI6MTAsImlucHV0Ijoia2V5IiwiZGF0YSI6e319XX19LCJwb3NpdGlvbiI6Wy02NzYuNzgwMTk0NTc5Mzg1NCw0MzMuMTY2OTMxMDkzMDg4MjRdLCJuYW1lIjoiQklQMzkifSwiMTQiOnsiaWQiOjE0LCJkYXRhIjp7Im1uZW1vbmljIjoiY2FtZXJhIHVuYWJsZSB0cmljayBpbmNvbWUgd2lzZG9tIG1peCByZW1pbmQgb3N0cmljaCBvZmZpY2UgbWFzdGVyIHN3YW1wIGxvdHRlcnkifSwiaW5wdXRzIjp7fSwib3V0cHV0cyI6eyJrZXkiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjE1LCJpbnB1dCI6ImtleSIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlstMTAyNi4wNzIwNjE4NDk4NzY3LDI2NS41NzE0OTg4ODQxOTIzXSwibmFtZSI6IkJJUDM5In0sIjE1Ijp7ImlkIjoxNSwiZGF0YSI6e30sImlucHV0cyI6eyJrZXkiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjE0LCJvdXRwdXQiOiJrZXkiLCJkYXRhIjp7fX1dfX0sIm91dHB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoxNiwiaW5wdXQiOiJwb2xpY2llcyIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlstMzYyLjQzMzExNjAxNTIxNDIsMjM3LjE3NDY1OTM5MzAyOTRdLCJuYW1lIjoiS2V5In0sIjE2Ijp7ImlkIjoxNiwiZGF0YSI6eyJ0aHJlc2giOjJ9LCJpbnB1dHMiOnsicG9saWNpZXMiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjEwLCJvdXRwdXQiOiJrZXkiLCJkYXRhIjp7fX0seyJub2RlIjoxNSwib3V0cHV0Ijoia2V5IiwiZGF0YSI6e319XX19LCJvdXRwdXRzIjp7InBvbCI6eyJjb25uZWN0aW9ucyI6W3sibm9kZSI6MTcsImlucHV0IjoicG9sMiIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlstNDIuNzg0Nzc3NTQ4NTc2NzQsMjUyLjIyMjU2NzM5MjgyMjAxXSwibmFtZSI6IlRocmVzaG9sZCJ9LCIxNyI6eyJpZCI6MTcsImRhdGEiOnt9LCJpbnB1dHMiOnsicG9sMSI6eyJjb25uZWN0aW9ucyI6W3sibm9kZSI6OCwib3V0cHV0IjoicG9sIiwiZGF0YSI6e319XX0sInBvbDIiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjE2LCJvdXRwdXQiOiJwb2wiLCJkYXRhIjp7fX1dfX0sIm91dHB1dHMiOnsicG9sIjp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjo2LCJpbnB1dCI6InBvbCIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlsyNTEuNjQ5NzA2MzcyNTA5MiwxMTEuMzMyODE0NzQ5MjcxNTVdLCJuYW1lIjoiQW5kIn19LCJjb21tZW50cyI6W3sidGV4dCI6IkFsaWNlIiwicG9zaXRpb24iOlstMTA4OS4zODA5NDUzOTI1OTM3LDIwMC4yMjM3Njc5NzM4MjYxN10sImxpbmtzIjpbMTRdLCJ0eXBlIjoiZnJhbWUiLCJ3aWR0aCI6MjgwLCJoZWlnaHQiOjQxNS45OTk5OTk5OTk5OTk5NH0seyJ0ZXh0IjoiQm9iIiwicG9zaXRpb24iOlstNzQzLjgyNTA0MTQzMzI4MjQsLTE1OS41NDIwOTc0MjI1MzU2OF0sImxpbmtzIjpbMl0sInR5cGUiOiJmcmFtZSIsIndpZHRoIjoyODAsImhlaWdodCI6NDE2fSx7InRleHQiOiJDYXJvbCIsInBvc2l0aW9uIjpbLTczOS4yMDk4NjkwODk1MTg0LDM2OS4yMDIxMjg0MjE0MTQxXSwibGlua3MiOlsxMl0sInR5cGUiOiJmcmFtZSIsIndpZHRoIjoyODAsImhlaWdodCI6NDE2fV0sIm5ldHdvcmsiOiJzaWduZXQifQ==

Reviewer 2: Should this link also include a line that connects Bob's key to the threshold (multisig) policy? Here's what it looks like with that line added: https://miniscript.fun/#/full/eyJpZCI6ImRlbW9AMC4xLjAiLCJub2RlcyI6eyIyIjp7ImlkIjoyLCJkYXRhIjp7Im1uZW1vbmljIjoiY3Jhc2ggZmF0YWwgaG9sbG93IHRoYW5rIHN3YWxsb3cgc3VibWl0IHRhdHRvbyBwb3J0aW9uIGNvZGUgZm9hbSBtYXRoIGZvcmNlIiwicGFzc3dvcmQiOiIiLCJkZXJpdmF0aW9uIjoiIn0sImlucHV0cyI6e30sIm91dHB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjozLCJpbnB1dCI6ImtleSIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlstNjgyLjk3ODA3NzAxNzExNTUsLTk0LjM1NDg3NDg2ODg2NzM4XSwibmFtZSI6IkJJUDM5In0sIjMiOnsiaWQiOjMsImRhdGEiOnt9LCJpbnB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoyLCJvdXRwdXQiOiJrZXkiLCJkYXRhIjp7fX1dfX0sIm91dHB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoxNiwiaW5wdXQiOiJwb2xpY2llcyIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlstMzU0LjM4NTI2MTYwMTMyMjYsLTI2LjE3MDA1ODk3OTU4NDMyMl0sIm5hbWUiOiJLZXkifSwiNiI6eyJpZCI6NiwiZGF0YSI6e30sImlucHV0cyI6eyJwb2wiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjE3LCJvdXRwdXQiOiJwb2wiLCJkYXRhIjp7fX1dfX0sIm91dHB1dHMiOnsiZGVzYyI6eyJjb25uZWN0aW9ucyI6W119fSwicG9zaXRpb24iOls1NTYuOTUxOTU3MTcxNjQxNCwxMjQuNjExNDg0NDU1Mzg4NjJdLCJuYW1lIjoiRGVzY3JpcHRvciJ9LCI4Ijp7ImlkIjo4LCJkYXRhIjp7Im51bSI6MjF9LCJpbnB1dHMiOnt9LCJvdXRwdXRzIjp7InBvbCI6eyJjb25uZWN0aW9ucyI6W3sibm9kZSI6MTcsImlucHV0IjoicG9sMSIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlstNzguNTU5ODUxODY0MzM4NDgsLTY1LjU0NzAzNTY2MTA1ODQ2XSwibmFtZSI6IkFmdGVyIn0sIjEwIjp7ImlkIjoxMCwiZGF0YSI6e30sImlucHV0cyI6eyJrZXkiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjEyLCJvdXRwdXQiOiJrZXkiLCJkYXRhIjp7fX1dfX0sIm91dHB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoxNiwiaW5wdXQiOiJwb2xpY2llcyIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlstMzQxLjIwMzA2MDUyMzAyNjMsNDk2LjUzNzY0OTU1NTYwNzRdLCJuYW1lIjoiS2V5In0sIjEyIjp7ImlkIjoxMiwiZGF0YSI6eyJtbmVtb25pYyI6InNwZWxsIGJvcmRlciBleGhhdXN0IHlhcmQgZWNvbm9teSBmb3NzaWwgYXJteSBndWVzcyB0eXBpY2FsIGJvbWIgYWNoaWV2ZSBhbW91bnQifSwiaW5wdXRzIjp7fSwib3V0cHV0cyI6eyJrZXkiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjEwLCJpbnB1dCI6ImtleSIsImRhdGEiOnt9fV19fSwicG9zaXRpb24iOlstNjc2Ljc4MDE5NDU3OTM4NTQsNDMzLjE2NjkzMTA5MzA4ODI0XSwibmFtZSI6IkJJUDM5In0sIjE0Ijp7ImlkIjoxNCwiZGF0YSI6eyJtbmVtb25pYyI6ImNhbWVyYSB1bmFibGUgdHJpY2sgaW5jb21lIHdpc2RvbSBtaXggcmVtaW5kIG9zdHJpY2ggb2ZmaWNlIG1hc3RlciBzd2FtcCBsb3R0ZXJ5In0sImlucHV0cyI6e30sIm91dHB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoxNSwiaW5wdXQiOiJrZXkiLCJkYXRhIjp7fX1dfX0sInBvc2l0aW9uIjpbLTEwMjYuMDcyMDYxODQ5ODc2NywyNjUuNTcxNDk4ODg0MTkyM10sIm5hbWUiOiJCSVAzOSJ9LCIxNSI6eyJpZCI6MTUsImRhdGEiOnt9LCJpbnB1dHMiOnsia2V5Ijp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoxNCwib3V0cHV0Ijoia2V5IiwiZGF0YSI6e319XX19LCJvdXRwdXRzIjp7ImtleSI6eyJjb25uZWN0aW9ucyI6W3sibm9kZSI6MTYsImlucHV0IjoicG9saWNpZXMiLCJkYXRhIjp7fX1dfX0sInBvc2l0aW9uIjpbLTM2Mi40MzMxMTYwMTUyMTQyLDIzNy4xNzQ2NTkzOTMwMjk0XSwibmFtZSI6IktleSJ9LCIxNiI6eyJpZCI6MTYsImRhdGEiOnsidGhyZXNoIjoyfSwiaW5wdXRzIjp7InBvbGljaWVzIjp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoxMCwib3V0cHV0Ijoia2V5IiwiZGF0YSI6e319LHsibm9kZSI6MTUsIm91dHB1dCI6ImtleSIsImRhdGEiOnt9fSx7Im5vZGUiOjMsIm91dHB1dCI6ImtleSIsImRhdGEiOnt9fV19fSwib3V0cHV0cyI6eyJwb2wiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjE3LCJpbnB1dCI6InBvbDIiLCJkYXRhIjp7fX1dfX0sInBvc2l0aW9uIjpbLTQyLjc4NDc3NzU0ODU3Njc0LDI1Mi4yMjI1NjczOTI4MjIwMV0sIm5hbWUiOiJUaHJlc2hvbGQifSwiMTciOnsiaWQiOjE3LCJkYXRhIjp7fSwiaW5wdXRzIjp7InBvbDEiOnsiY29ubmVjdGlvbnMiOlt7Im5vZGUiOjgsIm91dHB1dCI6InBvbCIsImRhdGEiOnt9fV19LCJwb2wyIjp7ImNvbm5lY3Rpb25zIjpbeyJub2RlIjoxNiwib3V0cHV0IjoicG9sIiwiZGF0YSI6e319XX19LCJvdXRwdXRzIjp7InBvbCI6eyJjb25uZWN0aW9ucyI6W3sibm9kZSI6NiwiaW5wdXQiOiJwb2wiLCJkYXRhIjp7fX1dfX0sInBvc2l0aW9uIjpbMjUxLjY0OTcwNjM3MjUwOTIsMTExLjMzMjgxNDc0OTI3MTU1XSwibmFtZSI6IkFuZCJ9fSwiY29tbWVudHMiOlt7InRleHQiOiJBbGljZSIsInBvc2l0aW9uIjpbLTEwODkuMzgwOTQ1MzkyNTkzNywyMDAuMjIzNzY3OTczODI2MTddLCJsaW5rcyI6WzE0XSwidHlwZSI6ImZyYW1lIiwid2lkdGgiOjI4MCwiaGVpZ2h0Ijo0MTUuOTk5OTk5OTk5OTk5OTR9LHsidGV4dCI6IkJvYiIsInBvc2l0aW9uIjpbLTc0My44MjUwNDE0MzMyODI0LC0xNTkuNTQyMDk3NDIyNTM1NjhdLCJsaW5rcyI6WzJdLCJ0eXBlIjoiZnJhbWUiLCJ3aWR0aCI6MjgwLCJoZWlnaHQiOjQxNn0seyJ0ZXh0IjoiQ2Fyb2wiLCJwb3NpdGlvbiI6Wy03MzkuMjA5ODY5MDg5NTE4NCwzNjkuMjAyMTI4NDIxNDE0MV0sImxpbmtzIjpbMTJdLCJ0eXBlIjoiZnJhbWUiLCJ3aWR0aCI6MjgwLCJoZWlnaHQiOjQxNn1dLCJuZXR3b3JrIjoic2lnbmV0In0=

Author: Yes, you are right! Which also makes the resulting output descriptor in the guide wrong.

Add the following to your `$DATA_DIR/bitcoin.conf`(as an alternative to using cli args):

```shell
debug=i2p # to get the relevant logs in your debug.log file
onlynet=i2p # to easily spot if there is a connectivity issue
i2psam=127.0.0.1:7656 # the usual address and port the SAM proxy is listening to
```

We will test this change by running an I2P node that accepts incoming connections and then a node that is not accepting incoming I2P connections. **Make sure that you are able to run an I2P router with SAM proxy enabled before proceeding with this test**.

> :bulb: Instead of scanning through the lengthy response of `getpeersinfo` in order to confirm your address for each outgoing connection, you can **use some `jq` magic to get a simplified view**:
>
> ```shell
> cli getpeerinfo | jq 'map(select(.network == "i2p" and .inbound == false)) | map({addrbind: .addrbind})'
> ```

First, run an I2P node that is accepting incoming connections and confirm that **your I2P address is the same for all outgoing connections**.

```shell
bitcoind -i2pacceptincoming=1
```

> 

Reviewer 3: As posted on the guide feedback issue  [26092](https://github.com/bitcoin/bitcoin/issues/26092) perhaps we could include this screenshot for better understanding of what to expect here:

![image](https://user-images.githubusercontent.com/110166421/195895485-cba24b04-6a91-44d8-b159-941cbbef3165.png)

Next, run an I2P node that is not accepting incoming connections and confirm that **your I2P address is different for each outgoing connection**.

```shell
bitcoind -i2pacceptincoming=0
```

> 

Reviewer 3: And here again as mentioned:

![image](https://user-images.githubusercontent.com/110166421/195894882-f6ace2e5-eb42-49cf-bb39-4148191fb182.png)
